haproxy/haproxy.go

```diff
@@ -183,14 +189,78 @@ func (h *HAproxy) WriteConfig(state *catalog.ServicesState, output io.Writer) er
 	return nil
 }
 
-// Execute a command and bubble up the error
+// notifySignals swallows a bunch of signals that get sent to us when running into
+// an error from HAproxy. If we didn't swallow these, the process would potentially
+// stop when the signals are propagated by the sub-shell.
+func (h *HAproxy) swallowSignals() {
+	// from HAproxy which propagate.
+	sigChan := make(chan os.Signal, 10)
+
+	// Used to stop the goroutine
+	h.sigStopChan = make(chan struct{})
+
+	go func() {
+		for {
+			select {
+			case <-sigChan:
+				// swallow signal
+			case <-h.sigStopChan:
+				break
+			}
+		}
+	}()
+
+	signal.Notify(sigChan, syscall.SIGSTOP, syscall.SIGTSTP, syscall.SIGTTIN, syscall.SIGTTOU)
+}
+
+// ResetSignals unhooks our signal handler from the signals the sub-commands
+// initiate. This is potentially destructive if other places in the program have
+// hooked to the same signals! Affected signals are SIGSTOP, SIGTSTP, SIGTTIN, SIGTTOU.
+func (h *HAproxy) ResetSignals() {
+	h.sigLock.Lock()
+	signal.Reset(syscall.SIGSTOP, syscall.SIGTSTP, syscall.SIGTTIN, syscall.SIGTTOU)
+	select {
+	case h.sigStopChan <- struct{}{}: // nothing
```

Comment:
Yep! If you have more than one goroutine waiting and you don't need to re-use the channel, you can call `close()` instead, which will unblock all of them.